### PR TITLE
Add support for fully localized chat event and additional support option

### PR DESCRIPTION
**Pull request by unDemian**
## Summary
This PR attempts to fix 1050-hg-gh by adding support for a new socket event `localized-support` based on which we would show an additional support option and track the customer preference. This solution is temporary to grab data and decide which option is better suited for our customers.

When customers try to reach us, if we don't have chat available in their native language, we would show them two options `Chat with us in English` or `Email us in your native language`. We will keep this for a couple of weeks to gather data about customer preferences.

**Tracks added**
`calypso_happychat_a_b_additional_support_option_shown` - When multiple support options are shown
`calypso_happychat_a_b_english_chat_selected` - When customers choose to have a chat in english
`calypso_happychat_a_b_native_ticket_selected` - When customers choose to submit a ticket in their native language

## Testing
1. Start Happychat with 1136-happychat-gh
2. Have an operator with support for `en`
3. Customer with locale `en` should have only the `Chat with us` option
4. Change customer locale to `es`
5. Customer should have the option to either chat with us in english or email us in `es`
6. Add operator suppport for `es`
7. Customer should again have only the chat with us option
8. Change customer's locale to `fr` or other not supported locale
9. The customer should have the option to either chat with us in `en` or email us in their native language

**Whenever the multiple support options are shown a track action should be dispatched for the event, also whenever the multiple support options are shown and one is selected a track action should be dispatched for the event.**

### CODE REVIEW

**Comment by unDemian** [comment]
Totally forgot to add the selector files 😄

File: client/state/happychat/selectors/has-happychat-localized-support.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ */
+import { get } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import isHappychatClientConnected from 'state/happychat/selectors/is-happychat-client-connected';
+
+/**
+ * Returns true if Happychat client is connected and server is available to take new chats
```
Comment (lamosty) [inline client/state/happychat/selectors/has-happychat-localized-support.js]: Shouldn't it say "... and server is available to take new localized chats"? This selector is only for localized support chats, right?

File: client/state/happychat/selectors/test/has-happychat-localized-support.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+				connection: {
+					status: HAPPYCHAT_CONNECTION_STATUS_UNINITIALIZED,
+					localizedSupport: true,
+				},
+			},
+		} );
+		expect( hasHappychatLocalizedSupport( state ) ).to.be.false;
+	} );
+
+	it( "should be false if Happychat isn't accepting new connections", () => {
```
Comment (lamosty) [inline client/state/happychat/selectors/test/has-happychat-localized-support.js]: I would also add here that it isn't accepting new localized chat connections or something like that. HC itself could be accepting connections but not for localized chats, or not?

File: client/state/happychat/connection/reducer.js [hunk 3]
```diff
@@ -65,8 +66,25 @@ const isAvailable = ( state = false, action ) => {
...
 
+/**
+ * Tracks whether happychat.io is accepting fully localized chats.
+ * Customers locale and HE locale should be an exact match ( doesn't default routing to english speaker HEs )
+ *
+ * @param  {Boolean} state  Current happychat status
+ * @param  {Object}  action Action playload
+ * @return {Boolean}        Updated happychat status
+ */
+const localizedSupport = ( state = false, action ) => {
```
Comment (lamosty) [inline client/state/happychat/connection/reducer.js]: Do you think we should also add tests for the reducer?

File: client/me/help/help-contact/index.jsx [hunk 4]
```diff
@@ -58,6 +59,9 @@ import {
 } from 'state/selectors';
 import QueryUserPurchases from 'components/data/query-user-purchases';
 import { getHelpSelectedSiteId } from 'state/help/selectors';
+import { getLanguage, isDefaultLocale } from 'lib/i18n-utils';
+const defaultLanguage = getLanguage( config( 'i18n_default_locale_slug' ) ).name;
```
Comment (lamosty) [inline client/me/help/help-contact/index.jsx]: Can we move this const into the Module variables block?

**Comment by lamosty** [comment]
Some things I noticed while testing:

- Normally, the button to send the support request through email is not enabled until a customer fills out the subject field. With this PR, the button is enabled but when clicked and subject field is not filled out, there's an error notice telling user to fill out the subject field.
- If you are ES and operator is EN only, you see the two buttons which is expected. Then, if you click to start chatting in English, you chat — also expected. If you reload Calypso and go to `/help/contact` again, you will see the two buttons again — expected. Now, if you start typing into the textarea, the button to chat disappears and only the button to send an email stays. When I look into HUD, I'm still assigned the one chat. Also, my throttle is at 3. The expected behavior to me would be to open the chat window after clicking to chat with me after that reload.

**Comment by unDemian** [comment]
Thanks for taking the time, I pushed some updates to address your feedback 🏅 

> Normally, the button to send the support request through email is not enabled until a customer fills out the subject field. With this PR, the button is enabled but when clicked and subject field is not filled out, there's an error notice telling user to fill out the subject field.

Good catch, fixed.

>If you are ES and operator is EN only, you see the two buttons which is expected. Then, if you click to start chatting in English, you chat — also expected. If you reload Calypso and go to /help/contact again, you will see the two buttons again — expected. Now, if you start typing into the textarea, the button to chat disappears and only the button to send an email stays. When I look into HUD, I'm still assigned the one chat. Also, my throttle is at 3. The expected behavior to me would be to open the chat window after clicking to chat with me after that reload.

Great catch, this happened because the track event `calypso_happychat_a_b_additional_support_option_shown` was recorded on each re-render of the contact form component. With track events being sent to happychat as websocket messages the customer got throttled.

File: client/me/help/help-contact/index.jsx (hunk 4 above)
Comment (unDemian) [reply client/me/help/help-contact/index.jsx]: Sure thing, good catch.

File: client/state/happychat/connection/reducer.js (hunk 3 above)
Comment (unDemian) [reply client/state/happychat/connection/reducer.js]: Definitely.

File: client/state/happychat/selectors/test/has-happychat-localized-support.js (hunk 2 above)
Comment (unDemian) [reply client/state/happychat/selectors/test/has-happychat-localized-support.js]: That is correct, updated.

File: client/state/happychat/selectors/has-happychat-localized-support.js (hunk 1 above)
Comment (unDemian) [reply client/state/happychat/selectors/has-happychat-localized-support.js]: That is correct, updated.

File: client/me/help/help-contact-form/index.jsx [hunk 5]
```diff
@@ -225,12 +226,36 @@ export class HelpContactForm extends React.PureComponent {
...
 	/**
 	 * Start a chat using the info set in state
 	 * @param  {object} event Event object
 	 */
 	submitForm = () => {
 		const { howCanWeHelp, howYouFeel, message, subject } = this.state;
+		const { additionalSupportOption } = this.props;
+
+		if ( additionalSupportOption && additionalSupportOption.enabled ) {
+			this.props.recordTracksEvent( 'calypso_happychat_a_b_english_chat_selected' );
```
Comment (mattwondra) [inline client/me/help/help-contact-form/index.jsx]: For this, and the other Tracks events, can you send additional props to note the customer's locale? That lets us also identify if there's a strong difference between `es` and `pt-br` speakers.

```
this.props.recordTracksEvent( 'calypso_happychat_a_b_english_chat_selected', { locale: customerLanguage } );
```

Comment (unDemian) [reply client/me/help/help-contact-form/index.jsx]: Great suggestion, added.

**Comment by mattwondra** [comment]
@unDemian This change has been noted internally, this is ready to merge first thing tomorrow (Friday).

**Pull request merged**